pkg/registry/core/service/storage/alloc.go

```diff
@@ -0,0 +1,1101 @@
+				klog.V(4).Infof("service %v/%v failed to clean up after failed service update error:%v. ShouldRelease/Released:%v/%v",
+					service.Namespace, service.Name, err, released, actuallyReleased)
+			}
+		},
+		revert: func() {
+			if dryRun {
+				return
+			}
+			if actuallyReleased, err := al.releaseIPs(allocated); err != nil {
+				klog.V(4).Infof("service %v/%v failed to clean up after failed service update error:%v. Allocated/Released:%v/%v",
+					service.Namespace, service.Name, err, allocated, actuallyReleased)
+			}
+		},
+	}
+	return txn, nil
+}
+
+// handles type change/upgrade/downgrade change type for an update service
+// this func does not perform actual release of clusterIPs. it returns
+// a map[family]ip for the caller to release when everything else has
+// executed successfully
+func (al *Allocators) updateClusterIPs(after After, before Before, dryRun bool) (allocated map[api.IPFamily]string, toRelease map[api.IPFamily]string, err error) {
+	oldService, service := before.Service, after.Service
+
+	// We don't want to auto-upgrade (add an IP) or downgrade (remove an IP)
+	// PreferDualStack services following a cluster change to/from
+	// dual-stackness.
+	//
+	// That means a PreferDualStack service will only be upgraded/downgraded
+	// when:
+	// - changing ipFamilyPolicy to "RequireDualStack" or "SingleStack" AND
+	// - adding or removing a secondary clusterIP or ipFamily
+	if isMatchingPreferDualStackClusterIPFields(after, before) {
+		return allocated, toRelease, nil // nothing more to do.
+	}
+
+	// use cases:
+	// A: service changing types from ExternalName TO ClusterIP types ==> allocate all new
+	// B: service changing types from ClusterIP types TO ExternalName ==> release all allocated
+	// C: Service upgrading to dual stack  ==> partial allocation
+	// D: service downgrading from dual stack ==> partial release
+
+	// CASE A:
+	// Update service from ExternalName to non-ExternalName, should initialize ClusterIP.
+	if oldService.Spec.Type == api.ServiceTypeExternalName && service.Spec.Type != api.ServiceTypeExternalName {
+		allocated, err := al.allocClusterIPs(service, dryRun)
+		return allocated, nil, err
+	}
+
+	// if headless service then we bail out early (no clusterIPs management needed)
+	if len(oldService.Spec.ClusterIPs) > 0 && oldService.Spec.ClusterIPs[0] == api.ClusterIPNone {
+		return nil, nil, nil
+	}
+
+	// CASE B:
+	// Update service from non-ExternalName to ExternalName, should release ClusterIP if exists.
+	if oldService.Spec.Type != api.ServiceTypeExternalName && service.Spec.Type == api.ServiceTypeExternalName {
+		toRelease = make(map[api.IPFamily]string)
+		if !utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
+			// for non dual stack enabled cluster we use clusterIPs
+			toRelease[al.defaultServiceIPFamily] = oldService.Spec.ClusterIP
+		} else {
+			// dual stack is enabled, collect ClusterIPs by families
+			for i, family := range oldService.Spec.IPFamilies {
```

Comment:
This is the bug that https://github.com/kubernetes/kubernetes/pull/104742 solves. The old service might be headless and this will yield into problems. Since the code is so different maybe we should just fix here as well (the fix PR has an integration test that will guard against this, but it is not merged yet so there is that)

Reply:
Please ignore this. it is fixed here: https://github.com/kubernetes/kubernetes/pull/96684/files/0dcc4e7193f744124c6afa650df0f5fb6921dfc8#diff-989117d719398a16288c815c08122ba373274bbfebc880bb78b67e405db21fd9R670